File: patpatbot/gpt.py
from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate


class Gpt:
    def __init__(self, openai_api_key, model="gpt-4o"):
        self.__model = ChatOpenAI(model=model, api_key=openai_api_key)
        self.__output_parser = StrOutputParser()

    def prompt(self, prompt_template_messages, prompt_data=None):
        """
        :param prompt_template_messages: List of tuples containing the message type and the message content.
            Example: [("system", "Your name is {name}."), ("human", "{user_input}")]
        :param prompt_data: Dictionary containing the data to be used in the prompt.
            Example: {"name": "TestBot", "user_input": "What is your name? Print **only** your name."}
        """
        if prompt_data is None:
            prompt_data = {}

        return self._do_prompt_from_messages(prompt_template_messages, prompt_data)

    def _do_prompt_from_messages(self, prompt_template_messages, prompt_data):
        prompt = ChatPromptTemplate.from_messages(prompt_template_messages)
        chain = prompt | self.__model | self.__output_parser
        return chain.invoke(prompt_data)

File: src/lib/Gpt.ts
import {ChatOpenAI} from '@langchain/openai';
import {ChatPromptTemplate} from '@langchain/core/prompts';
import {StringOutputParser} from '@langchain/core/output_parsers';
import {IQueryable, FlatObject} from "./types";
import logger from "./logging";
import PromptTemplate from "./PromptTemplate";

type PromptTemplateMessages = Array<['system' | 'human' | 'ai', string]>;

const PROMPT_SYSTEM_DEFAULT = "You are a senior technical writer.";

class Gpt implements IQueryable {
    private model: ChatOpenAI;
    private promptTemplateMessages: PromptTemplateMessages = [["system", PROMPT_SYSTEM_DEFAULT]];

    constructor(openaiApiKey: string, model: string = "gpt-4o") {
        this.model = new ChatOpenAI({ model, apiKey: openaiApiKey });
    }

    reset() {
        this.promptTemplateMessages = [["system", PROMPT_SYSTEM_DEFAULT]];
    }

    async execute(
        promptTemplate: PromptTemplate,
        promptData: FlatObject = {},
    ): Promise<FlatObject> {
        if (promptTemplate.promptSystem) {
            this.promptTemplateMessages[0] = ["system", promptTemplate.promptSystem];
        }
        this.promptTemplateMessages.push(["human", promptTemplate.promptHuman]);
        const output = await this.doPromptFromMessages(promptData);
        this.promptTemplateMessages.push(["ai", this.escapeCurlyBraces(output)]);
        logger.info(`Last AI reply: %o [...]`, output.slice(0, 100));

        return promptTemplate.formatOutput(output);
    }

    private async doPromptFromMessages(promptData: FlatObject): Promise<string> {
        return await ChatPromptTemplate
            .fromMessages(this.promptTemplateMessages, {})
            .pipe(this.model)
            .pipe(new StringOutputParser())
            .invoke(promptData);
    }

    /**
     * Escapes curly braces in the string to avoid confusing the template engine
     * with GPT responses containing curly braces.
     */
    private escapeCurlyBraces(str: string): string {
        return str
            .replace(/\{/g, '&#x7B;')
            .replace(/}/g, '&#x7D;');
    }
}

export default Gpt;
